src/wally-connector.ts

```diff
@@ -325,6 +397,33 @@ class WallyConnector {
     return res;
   }
 
+  /**
+   * The promise for handling when trying to make a request before the user has
+   * logged in. Either:
+   * - trigger a login once (web3 standard), and trigger the request after the
+   *   login is complete (adding requests in the meantime to the emitter queue) OR
+   * - just add all requests to the emitter queue, waiting for the consumer to manually login.
+   * @param req RequestObj
+   * @returns Promise
+   */
+  private deferredRequest<T extends MethodNameType>(
+    req: RequestObj<T>
+  ): Promise<MethodResponse<T> | null> {
+    return new Promise((resolve, reject) => {
+      if (!this.disableLoginOnRequest && !this.isLoggingIn) {
+        this.login().then(() => {
+          resolve(this.request(req));
+        });
```

Comment:
Q: Why not use `async` and `await`? Seeing a lot of code that constructs promises.

Reply:
left a comment above - to me it's more explicit than relying on `await` + `return`, which in my experience has sometimes not worked out the way i intended. i _think_ there might also still be some iffy support.

what's the advantage of using async/await?

Reply:
Either or is fine. Most times the async and await will come out cleaner. async-await and promises are pretty much identical except one is more procedural and arguably a little more "OO-ish". In order to achieve the above, you'd put the function together with await and then after that do the emitting if they're supposed to be done together so something like the following. 

```javascript
async function myFunc() {
  await login();
  emit(...);
}

// in the main caller
await myFunc();
```

Reply:
oh yeah i mean i'm definitely familiar with how async/await works, i was just commenting on how it gets a little bit more complicated when you're resolving promises inside of other callbacks - and I don't think it maps very cleanly:
```js
return new Promise((resolve, reject) => {
  const listener = () => {
    this.removeListener(EmitterMessage.ACCOUNTS_CHANGED, listener);
    resolve(this.request(req));
  };
  this.on(EmitterMessage.ACCOUNTS_CHANGED, listener);
})
```
tbh i don't know if i'm comfortable refactoring this logic just for the sake of being a little cleaner

Reply:
While a promise can have the behavior above, it's actually a little hacky because say if you `await` the promise that is returned but the callback is never triggered, then it would block. It prevents flexibility in the code, because then a promise always needs to be used for every single scenario and if it doesn't resolve, the chaining of the promises would never resolve. 

What you want instead is probably a callback function to passed into the function and decorate that function with the remove listener and pass the return value into that same callback.